# Integral of sin(x)+cos(x) from 0.4 raidans to 1.6 raidans

In [6]:
import math
import time

def trapezoidalRule(function,a,b):
    start_time = time.time()
    h = b-a
    x0 = a
    x1 = b
    d = (h/2)*(function(x0)+function(x1))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def simpsonRule(function,a,b):
    start_time = time.time()
    h = (b-a)/2
    x0 = a
    x1 = a+h
    x2 = b
    d = (h/3)*(function(x0)+(4*function(x1))+function(x2))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def booleRule(function,a,b):
    start_time = time.time()
    h = (b-a)/4
    x0 = a
    x1 = a+h
    x2 = x1+h
    x3 = x2+h
    x4 = b
    d = ((2*h)/45)*((7*function(x0))+(32*function(x1))+(12*function(x2))+(32*function(x3))+(7*function(x4)))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def compositeSimpsonRule(function,a,b):
    start_time = time.time()
    n = 1000000
    h = (b-a)/n
    x = [a+(j*h) for j in range(0,n+1)]
    secondterm = 0
    for j in range(1,int(n/2)):
        secondterm += function(x[(2*j)-1])
    thirdterm = 0
    for j in range(1,int(n/2)+1):
        thirdterm += function(x[2*j])
    d = (h/3)*((function(a))+(2*secondterm)+(4*thirdterm)+(function(b)))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def compositeTrapezoidalRule(function,a,b):
    start_time = time.time()
    n = 1000000
    h = (b-a)/n
    x = [a+(j*h) for j in range(0,n+1)]
    secondterm = 0
    for j in range(1,n):
        secondterm += function(x[j])
    d = (h/2)*((function(a))+(2*secondterm)+(function(b)))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def compositeMidpointRule(function,a,b):
    start_time = time.time()
    n = 1000000
    h = (b-a)/n
    x = [a+(j*h) for j in range(0,n+1)]
    secondterm = 0
    for j in range(2,int(n/2)+1):
        secondterm += function(x[2*j])
    d = (2*h)*(secondterm)
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def S(function,a,b):
    h = (b-a)/2
    return ((h/3)*(function(a)+(4*function(a+h))+function(b)))


def adaptiveQuadrature(function,a,b):
    start_time = time.time()
    firstterm = S(function,a,((a+b)/2))
    secondterm = S(function,((a+b)/2),b)
    thirdterm = (1/16)*((16/15)*((S(function,a,b))-(S(function,a,((a+b)/2)))-(S(function,((a+b)/2),b))))
    d = firstterm+secondterm-thirdterm
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time


def main():
    f = lambda x: math.sin(x) + math.cos(x)
    a = 0.4
    b = 1.6
    exact = (math.sin(1.6)-math.cos(1.6))-(math.sin(0.4)-math.cos(0.4))
    trapezoidalrule, trapezoidalruletime = trapezoidalRule(f,a,b)
    simpsonrule, simpsonruletime = simpsonRule(f,a,b)
    boolerule, booleruletime = booleRule(f,a,b)
    compositetrapezoidal, compositetrapezoidaltime = compositeTrapezoidalRule(f,a,b)
    compositesimpson, compositesimpsontime = compositeSimpsonRule(f,a,b)
    compositemidpointrule, compositemidpointruletime = compositeMidpointRule(f,a,b)
    adaptivequadrature, adaptivequadraturetime = adaptiveQuadrature(f,a,b)
    print(f"Trapezoidal        \tValue: {trapezoidalrule}\tTime: {trapezoidalruletime}")
    print(f"Simpson            \tValue: {simpsonrule}\tTime: {simpsonruletime}")
    print(f"Boole              \tValue: {boolerule}\tTime: {booleruletime}")
    print(f"Adaptive Quadrature\tValue: {adaptivequadrature}\tTime: {adaptivequadraturetime}")
    print(f"Composite Midpoint \tValue: {compositemidpointrule}\tTime: {compositemidpointruletime}")
    print(f"Composite Simpson  \tValue: {compositesimpson}\tTime: {compositesimpsontime}")
    print(f"Composite Trapezoid\tValue: {compositetrapezoidal}\tTime: {compositetrapezoidaltime}")
    print(f"Exact              \tValue: {exact}")
    
main()

Trapezoidal        	Value: 1.3685120502310513	Time: 2.86102294921875e-06
Simpson            	Value: 1.5615893159511798	Time: 2.384185791015625e-06
Boole              	Value: 1.5604132512202857	Time: 3.0994415283203125e-06
Adaptive Quadrature	Value: 1.5604132512202857	Time: 1.1920928955078125e-05
Composite Midpoint 	Value: 1.5604122237565325	Time: 0.2349236011505127
Composite Simpson  	Value: 1.560415640993598	Time: 0.4228219985961914
Composite Trapezoid	Value: 1.560415777036792	Time: 0.3605208396911621
Exact              	Value: 1.5604157770370284


<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">Trapezoidal Rule</th>
    <th class="tg-baqh">Simpson's Rule</th>
    <th class="tg-baqh">Boole's Rule</th>
    <th class="tg-baqh">Composite Trapezoidal Rule</th>
    <th class="tg-baqh">Composite Midpoint Rule</th>
    <th class="tg-baqh">Composite Simpson Rule</th>
    <th class="tg-baqh">Adaptive Quadrature</th>
  </tr>
  <tr>
    <td class="tg-baqh">Values</td>
    <td class="tg-baqh">1.3685120502310513</td>
    <td class="tg-baqh">1.5615893159511798</td>
    <td class="tg-baqh">1.5604132512202857</td>
    <td class="tg-baqh">1.560415777036792</td>
    <td class="tg-baqh">1.5604122237565325</td>
    <td class="tg-baqh">1.560415640993598</td>
    <td class="tg-baqh">1.5604132512202857</td>
  </tr>
  <tr>
    <td class="tg-baqh">Exact Value</td>
    <td class="tg-baqh">1.5604157770370284</td>
    <td class="tg-baqh">1.5604157770370284</td>
    <td class="tg-baqh">1.5604157770370284</td>
    <td class="tg-baqh">1.5604157770370284</td>
    <td class="tg-baqh">1.5604157770370284</td>
    <td class="tg-baqh">1.5604157770370284</td>
    <td class="tg-baqh">1.5604157770370284</td>
  </tr>
  <tr>
    <td class="tg-baqh">Times</td>
    <td class="tg-baqh">9.059906005859375e-06</td>
    <td class="tg-baqh">7.62939453125e-06</td>
    <td class="tg-baqh">9.775161743164062e-06</td>
    <td class="tg-baqh">0.4296834468841553</td>
    <td class="tg-baqh">0.24765372276306152</td>
    <td class="tg-baqh">0.43339085578918457</td>
    <td class="tg-baqh">1.2874603271484375e-05</td>
  </tr>
  <tr>
    <td class="tg-baqh">Absolute Error</td>
    <td class="tg-baqh">0.19190372680597712</td>
    <td class="tg-baqh">-0.0011735389141513952</td>
    <td class="tg-baqh">2.5258167426489564e-06</td>
    <td class="tg-baqh">2.3647750424515834e-13</td>
    <td class="tg-baqh">3.5532804958382513e-06</td>
    <td class="tg-baqh">1.360434302988267e-07</td>
    <td class="tg-baqh">2.5258167426489564e-06</td>
  </tr>
  <tr>
    <td class="tg-baqh">Relative Error</td>
    <td class="tg-baqh">14.02280139028204</td>
    <td class="tg-baqh">0.07515029093526943</td>
    <td class="tg-baqh">0.00016186844995540116</td>
    <td class="tg-baqh">1.5154775267283305e-11</td>
    <td class="tg-baqh">0.00022771421818806907</td>
    <td class="tg-baqh">8.718409808568745e-06</td>
    <td class="tg-baqh">0.00016186844995540116</td>
  </tr>
</table>

# Discussion/Analysis

Based on the results of this experiment, it can be seen that there is a correlation between the time it takes to calculate the integral and the accuracy of the result. More notably, the methods that involve summation (Composite Midpoint Rule, Composite Trapezoidal Rule, and Composite Simpson Rule) take considerably longer to compute as opposed to the methods that use a direct formula (Trapezoidal Rule, Simpson's Rule, Boole's Rule, and Adaptive Quadrature). Surprisingly, Boole's Rule and the Adaptive Quadrature Method result in the same outcome but Boole's Rule is faster by one degree of magnitude. Boole's Rule also has the lowest relative error out of the non-summation based methods tested, making it the best non-summation method to use. In terms of the methods that use summation, the number of subdivisions used, n, is kept constant between the three functions. In this case, n is 1000000. All three methods were within one-tenth of a second in completion time. However, the relative error of the Composite Trapezoidal Rule was considerably smaller than the other two methods, with a difference of five orders of magnitude for the Composite Simpson Rule and seven orders of magnitude for the Composite Midpoint Rule. Overall, the errors for the various methods are aligned with the expected errors.